docs/metrics-data-model.md

```diff
@@ -0,0 +1,177 @@
+# k6 Time Series
+
+We would migrate to a time series data model for getting the benefits of aggregated data so just allocating samples one time and pass around the aggregated time series value (or values - e.g. in case of Trend). The Outputs will switch from fetching the buffered samples to fetching a snapshot of the time series (the same concept adopted by Prometheus' scraping operation).
```

Comment:
ok, I will try to rewrite this part. If we exclude the aggregation from the time series concept what are the other benefits we get?

I see:

* Efficient comparison - we can compare quickly if a time series matches another just by checking the hash.
* Memory storage - we can reference the hash instead to pass the entire name+tags where tags will end in a copy
* Post-run analysis  - We can count the number of generated time series so we know the unique combinations
* Do we have more?
 
This should be also a quick summary of what `1831` is already documenting.

>For example, this output will benefit from both, since Prometheus deals with aggregated data. However, the JSON and CSV outputs certainly wont. The most they will benefit from the grouping of metric samples in time series is to maybe cache some things a bit better. The end-of-test summary will benefit somewhat, but not its aggregation spans multiple time series.

Does it mean we will continue to flush all the samples to the outputs then each output will sink/aggregate autonomously?

Reply:
I have a feeling that we converge on the idea that we don't need time series we just want `SampleTags` to be comparable in `O(1)` time? If you want to call this time series, I am fine with that as well. 

And then we will be aggregating in each output? With the possibility of there being some in core place where we also aggregate them in some manner ... possible in a future version?

Reply:
Kind of... :thinking: If `SampleTags` are comparable in constant time, then they can be used as an index. So that gives us the ability to efficiently build a time series for every unique group of metric tags we want :man_shrugging: But we don't always need the exact same groupings, and we don't always need aggregation. So, by necessity, these things would have to be decoupled.

Reply:
>If SampleTags are comparable in constant time

Just to verify my alignment on this, in the solution in the document we can't really compare `SampleTags` in O(1), we can verify it in O(N). What we can compare in O(1) is the name+SampleTags that in the end most of the time will be the same thing.

But there are exceptions, for example, in the Ingester the first time it sees a new time series it should create a tree of all the parent's time series to know the full set of Sinks to invoke and we have to check all the single tags in the SampleTags for getting the right tree. WDYT? (I have the feeling I'm missing potential limits of this proposal about the future features).

Based on it and that we want distributed sinks, for now, do we really need to involve the name in the hash? In this way, we could avoid to refactor the signature here https://github.com/grafana/k6/blob/556747fbe585c981507904d634e705c68891e663/metrics/engine/ingester.go#L83-L89 mentioned in the other comment.

In the case of sinks per time series, we could have a nested index (e.g `map[metricName]map[SampleTagHash]*Sink`).

Reply:
> Just to verify my alignment on this, in the solution in the document we can't really compare SampleTags in O(1), we can compare it in O(N). What we can compare in O(1) is the name+SampleTags that in the end most of the time will be the same thing.

Yeah, sorry, that's what I meant. If the `metric name + tag set`, i.e. the "time series", is comparable in constant time, we can use it as an index.

> But there are exceptions, for example, [in the Ingester](https://github.com/grafana/k6/blob/556747fbe585c981507904d634e705c68891e663/metrics/engine/ingester.go#L77-L92) the first time it sees a new time series it should create a tree of all the parent's time series to know the full set of Sinks to invoke and we have to check all the single tags in the SampleTags for getting the right tree. Right?

Kind of. When the `MetricsEngine`'s ingester sees a `Sample` with a new (for it) time series, it will need to see which of its own sinks for the `Sample`'s `Metric` match the tags of the new time series. And yeah, that operation is not constant, it's actually `O(m*n)`, where `m` is the number sub-metrics (sub-metric thresholds) and `n` is the length of the constraints for each. However, the improvement here is that this matching `Sink` list (not a tree, just a simple slice) can be cached for the next time a metric from the same time series is encountered! 

> Based on it and that we want distributed sinks, for now, do we really need to involve the name in the hash? In this way, we could avoid refactoring the signature here https://github.com/grafana/k6/blob/556747fbe585c981507904d634e705c68891e663/metrics/sample.go#L292-L318 mentioned in the other comment. WDYT? (I have the feeling I'm missing the potential limits of this proposal regarding the future features).
> 
> In the case of sinks per time series, we could have a nested index (e.g map[metricName]map[SampleTagHash]*Sink)

Hmm, interesting idea, I am not sure which will be better... :thinking: 

On the one hand, if we don't refactor it, we might introduce more complexity everywhere else :thinking: We'd need either these nested maps or composite indexes with `(metric_name, tag_set_hash)` tuples everywhere, right? Even if we ignore potential comparison and indexing inefficiencies, that will add development complexity, I think?

On the other hand, if we don't include the metric name in the hash, we'd be able to potentially more easily correlate Samples from different metrics but with the same tags :thinking: Not sure if that would be useful for anything, since we already package related samples in SampleContainers, but :man_shrugging: 

I think the first (including the metric name in the hash) is the better way to go here, but I am far from sure.